### ISSUE TITLE
Do-not-evict pods being evicted

### ISSUE THREAD

**Issue by theMagicalKarp** [aws/karpenter-core#588]
**Observed Behavior**:
We've identified an issue with Karpenter, where it's evicting pods that have been labeled with the `karpenter.sh/do-not-evict` annotation. The problem is evident in the following event:

```
Evicted: Evicted pod
```
Which was attached to a pod with a `karpenter.sh/do-not-evict: "true"` annotation.

Upon examining the Karpenter logs, we found a relevant log entry that coincided with these improper evictions, with near-exact timestamps.

```
deprovisioning via consolidation delete, terminating 1 machines xxxxxxxxxxxxxxxxxxxxx
```

**Expected Behavior**:
Pods with the `karpenter.sh/do-not-evict: "true"` annotation should not be forcibly evicted.

**Reproduction Steps** (Please include YAML):
Our environment runs a cluster that dynamically scales to ~100 nodes, accommodating ephemeral workloads scheduled in large bursts, often completing in just 2-5 minutes. We believe this rapid scaling introduces a lot of pressure on Karpenter.

This is what our provisioner looks like.
```YAML
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  consolidation:
    enabled: true
  limits:
    resources:
      cpu: 5k
  providerRef:
    name: default
  requirements:
  - key: karpenter.k8s.aws/instance-category
    operator: In
    values:
    - c
    - m
    - r
  - key: karpenter.k8s.aws/instance-cpu
    operator: Gt
    values:
    - "15"
  - key: karpenter.k8s.aws/instance-cpu
    operator: Lt
    values:
    - "33"
  - key: karpenter.k8s.aws/instance-hypervisor
    operator: In
    values:
    - nitro
  - key: karpenter.k8s.aws/instance-generation
    operator: Gt
    values:
    - "4"
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - spot
    - on-demand
  - key: kubernetes.io/os
    operator: In
    values:
    - linux
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  ttlSecondsUntilExpired: 86400
```

**Versions**:
- Chart Version: `v0.29.2`
- Kubernetes Version (`kubectl version`): `Server Version: v1.24.16-eks-2d98532`

## Concern
The question I'm raising is whether the following section of code should also consider the `do-not-evict` annotation. It seems plausible that a race condition may arise where nodes are selected for consolidation, but just before consolidation, a workflow is scheduled, even before cordoning the node.

https://github.com/aws/karpenter-core/blob/8549d55693b3cc3acd3b35284334b40bbd33cf30/pkg/controllers/termination/terminator/eviction.go#L119-L141

**Comment by jonathan-innis** [comment]
Duplicate of this issue can be found in `aws/karpenter`: https://github.com/aws/karpenter/issues/4310. This issue was resolved with #583 and will be released with a patch fix shortly.

**Comment by theMagicalKarp** [comment]
Sweet, thanks! I'll keep an eye out for that release then.
